### PR TITLE
Add block merkle proof to batch justification

### PR DESCRIPTION
**Pull request by nomaxg**
Modifies the batcher to append a (now mocked) espresso block merkle proof to the batch justification.

### CODE REVIEW

**Comment by ImJeremyHe** [comment]
I don't know well about the light client. From the changes, it seems that the sequencer can not get the `proof` and `raw transaction` simultaneously. That's why we need to add the proof to the L2 message in the batch poster.
If then, I prefer to remove the justification part from the L2 message. Poster, validator can just access the proof when they need it. I think it is unnecessary to store the proof into the messages.

**Comment by nomaxg** [comment]
> I don't know well about the light client. From the changes, it seems that the sequencer can not get the `proof` and `raw transaction` simultaneously. That's why we need to add the proof to the L2 message in the batch poster. If then, I prefer to remove the justification part from the L2 message. Poster, validator can just access the proof when they need it. I think it may be unnecessary to store the proof into the messages.

The sequencer can get both, but we figured it would be better to not block the sequencer (this would disable "preconfirmations"). The proof needs to make it to the L1 somehow, because we are still at a point where we require L1 data availability (i.e, everything required to verify the L2 chain exists on the L1). If we post a batch without a proof, we would be violating this requirement.

**Comment by nomaxg** [comment]
> I don't know well about the light client. From the changes, it seems that the sequencer can not get the `proof` and `raw transaction` simultaneously. That's why we need to add the proof to the L2 message in the batch poster. If then, I prefer to remove the justification part from the L2 message. Poster, validator can just access the proof when they need it. I think it may be unnecessary to store the proof into the messages.

The sequencer can get both, but we figured it would be better to not block the sequencer (this would disable "preconfirmations"). The proof needs to make it to the L1 somehow, because we are still at a point where we require L1 data availability (i.e, everything required to verify the L2 chain exists on the L1). If we post a batch without a proof, we would be violating this requirement.

File: arbnode/batch_poster.go [hunk 1]
```diff
@@ -427,6 +463,39 @@ func AccessList(opts *AccessListOpts) types.AccessList {
 	return l
 }
 
+// Adds a block merkle proof to an Espresso justification, providing a proof that a set of transactions
+// hashes to some light client state root
+func (b *BatchPoster) addEspressoBlockMerkleProof(
+	msg *arbostypes.MessageWithMetadata,
+) (*arbostypes.MessageWithMetadata, error) {
+	if msg.Message.Header.Kind == arbostypes.L1MessageType_L2Message &&
+		msg.Message.L2msg[0] == arbos.L2MessageKind_EspressoTx {
```
Comment (sveitser) [inline arbnode/batch_poster.go:472]: This check happens in multiple places now, we could consider extracting a helper function for it.

File: arbnode/batch_poster.go [hunk 2]
```diff
@@ -427,6 +463,39 @@ func AccessList(opts *AccessListOpts) types.AccessList {
 	return l
 }
 
+// Adds a block merkle proof to an Espresso justification, providing a proof that a set of transactions
+// hashes to some light client state root
+func (b *BatchPoster) addEspressoBlockMerkleProof(
+	msg *arbostypes.MessageWithMetadata,
+) (*arbostypes.MessageWithMetadata, error) {
```
Comment (sveitser) [inline arbnode/batch_poster.go:470]: I think this function doesn't really need to return `msg` because it modifies it.

**Review by sveitser — APPROVED** [review]
LGTM just some minor comments.

File: arbnode/transaction_streamer.go [hunk 3]
```diff
@@ -700,6 +701,26 @@ func (s *TransactionStreamer) addMessagesAndEndBatchImpl(messageStartPos arbutil
 	var hasNewConfirmedMessages bool
 	var cacheClearLen int
 
+	for i := 0; i < len(messages); i++ {
```
[the remark this answers is not in the document]
Comment (nomaxg) [reply arbnode/transaction_streamer.go:704]: I think that might end up being the approach longer term - I'll add a comment here, agree that it is a bit risky

File: system_tests/espresso_e2e_test.go [hunk 4]
```diff
@@ -142,7 +144,7 @@ func createValidationNode(ctx context.Context, t *testing.T, jit bool) func() {
 
 }
 
-func createL1ValidatorPosterNode(ctx context.Context, t *testing.T) (*NodeBuilder, func()) {
+func createL1ValidatorPosterNode(ctx context.Context, t *testing.T, hotshotUrl string) (*NodeBuilder, func()) {
```
[the remark this answers is not in the document]
Comment (nomaxg) [reply system_tests/espresso_e2e_test.go:147]: we still do to instantiate the mock client

**Pull request merged**